profiles/sam/terraform/sam/sam.tf

```diff
@@ -1,87 +1,28 @@
+provider "google" {
+  alias = "dns"
+  project     = "${var.dns_project}"
+  region      = "${var.dns_region}"
+  credentials = "${file("dns_sa.json")}"
+}
```

Comment:
why is the DNS block different between sam.tf and thurloe.tf? I expect that after this PR, the only diff between those two files should be the cloudsql code.

Reply:
I'm going to answer the question that I think you're asking, but if I'm wrong about what that is let me know and I'll try again. 

This block is the terraform provider for the DNS records. The provider is the bundle of credentials-plus-account-details that tells terraform where to make the resources.

When it comes to DNS, we have three options that I know of:

1. For each environment that we want, 
  a) create a DNS zone for `<env>.broadinstitute.org`
  b) get BITS to delegate the DNS for `<env>.broadinstitute.org` to the DNS servers associated with the zone we made
  c) go to a CA and do the certificate-purchasing dance, get the CA, 
  d) put the certificate in vault
  e) use that certificate in the render & deploy phases
2) In each perf project
  a) create a DNS zone with a static name, such as `dsde-qa.broadinstitute.org`, which we already have a certificate for
  b) use the certificate associated with that zone
  c) Since that DNS zone will not be "official" (i.e. _every_ perf env has it's own `dsde-qa.broadinstitute.org` DNS zone and none of them are the officially delegated one that your DNS would actually resolve to), in every host and every instance, manually or programmatically override the DNS settings so that the host uses the DNS servers provided by the environment.
3) Start from a DNS zone that exists and for which a certificate exists--in this case `dsde-perf.broadinstitute.org`
  a) each environment, when it needs to create a DNS record, creates a (unique across all environments) DNS record in the `dsde-perf.broadinstitute.org` zone, _which is managed  in the broad-dsde-perf google project_, and uses certificates for that zone from Vault.

I strongly favor option 3. What that means is that the `load-balanced-instances` module needs two providers: one with the creds and project / region etc to create the load balancer and other resources in the perf project, and the other to create the DNS records in the DNS project (`broad-dsde-perf`). When you are in this situation, the best practice in terraform is to [always define the providers in the root module](https://www.terraform.io/docs/configuration/modules.html#providers-within-modules). That practice allows more flexibility at this level of the hierarchy.